tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/NativeScope.java

```diff
@@ -0,0 +1,145 @@
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================
+*/
+package org.tensorflow.op;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import org.tensorflow.DeviceSpec;
+import org.tensorflow.ExecutionEnvironment;
+import org.tensorflow.Graph;
+import org.tensorflow.GraphOperation;
+import org.tensorflow.Operation;
+import org.tensorflow.OperationBuilder;
+import org.tensorflow.internal.c_api.TF_Operation;
+import org.tensorflow.internal.c_api.TF_Scope;
+
+import static org.tensorflow.internal.c_api.global.tensorflow.*;
+
+/** A {@link Scope} implementation backed by a native scope. */
+public final class NativeScope implements Scope {
+
+  @Override
+  public ExecutionEnvironment env() {
+    return graph;
+  }
+
+  @Override
+  public NativeScope withSubScope(String childScopeName) {
+    return new NativeScope(TF_NewSubScope(nativeScope, childScopeName), graph, null, device);
+  }
+
+  @Override
+  public NativeScope withName(String opName) {
+    return new NativeScope(nativeScope, graph, opName, device);
+  }
+
+  @Override
+  public NativeScope withNameAsSubScope(String defaultName) {
+    if (opName == null) {
+      return withSubScope(defaultName);
+    } else {
+      return withSubScope(opName);
+    }
+  }
+
+  @Override
+  public NativeScope withDevice(DeviceSpec newDevice) {
+    return new NativeScope(
+        TF_NewScopeWithDevice(nativeScope, newDevice.toString()), graph, newDevice.toString());
+  }
+
+  @Override
+  public Scope withInitScope() {
+    throw new IllegalStateException("Can't add init operations in a gradient scope");
+  }
+
+  @Override
+  public String makeOpName(String defaultName) {
+    String name = opName != null ? opName : defaultName;
+    return TF_GetUniqueNameForOp(nativeScope, name);
+  }
+
+  @Override
+  public String makeUnique(String id) {
+    return TF_GetUniqueNameForOp(nativeScope, id);
+  }
+
+  @Override
+  public void refreshNames() {}
+
+  @Override
+  public Scope withControlDependencies(Iterable<Op> controls) {
+    return withControlDependencyOps(StreamSupport.stream(controls.spliterator(), false)
+            .map(Op::op)
+            .collect(Collectors.toList()));
+  }
+
+  @Override
+  public Scope withControlDependencyOps(Iterable<Operation> controls) {
+    List<Operation> controlDeps =
+        StreamSupport.stream(controls.spliterator(), false).collect(Collectors.toList());
```

Comment:
This juggling feels annoying given it's exactly the same juggling a few lines above. Maybe we should change the API (in a different PR) to accept a `List`?

Reply:
Yes, I did a few "optimization" around but tried to keep the Java implementation as much as it was before in this PR.